contracts/modules/elrond-wasm-module-governance/src/lib.rs

```diff
@@ -0,0 +1,411 @@
+#![no_std]
+
+elrond_wasm::imports!();
+
+pub mod governance_configurable;
+
+pub mod governance_proposal;
+use governance_proposal::*;
+
+#[elrond_wasm_derive::module]
+pub trait GovernanceModule:
+	governance_configurable::GovernanceConfigurablePropertiesModule
+{
+	// TODO: Withdraw functionality after voting ends
+
+	// endpoints
+
+	// Used to deposit tokens for "payable" actions
+	// There is no "withdraw" functionality
+	// Funds can only be retrived through an action
+	#[payable("*")]
+	#[endpoint(depositTokensForAction)]
+	fn deposit_tokens_for_action(&self) {}
+
+	// Used to withdraw the tokens after the action was executed or cancelled
+	#[endpoint(withdrawGovernanceTokens)]
+	fn withdraw_governance_tokens(&self, proposal_id: usize) -> SCResult<()> {
+		self.require_valid_proposal_id(proposal_id)?;
+		require!(
+			self.get_proposal_status(proposal_id) == GovernanceProposalStatus::None,
+			"Proposal has to be executed or canceled first"
+		);
+
+		let caller = self.blockchain().get_caller();
+		let governance_token_id = self.governance_token_id().get();
+		let nr_votes_tokens = self.votes(proposal_id).get(&caller).unwrap_or_default();
+		let nr_downvotes_tokens = self.downvotes(proposal_id).get(&caller).unwrap_or_default();
+
+		if nr_votes_tokens > 0 {
+			self.send()
+				.direct(&caller, &governance_token_id, &nr_votes_tokens, &[]);
+
+			self.votes(proposal_id).remove(&caller);
+		}
+		if nr_downvotes_tokens > 0 {
+			self.send()
+				.direct(&caller, &governance_token_id, &nr_downvotes_tokens, &[]);
+
+			self.downvotes(proposal_id).remove(&caller);
+		}
+
+		Ok(())
+	}
+
+	#[payable("*")]
+	#[endpoint]
+	fn propose(
+		&self,
+		#[payment_amount] payment_amount: Self::BigUint,
+		description: BoxedBytes,
+		#[var_args] actions: VarArgs<GovernanceActionAsMultiArg<Self::BigUint>>,
+	) -> SCResult<usize> {
+		self.require_payment_token_governance_token()?;
```

Comment:
For consistency I think it's best that you either take payment info from #payment parameters or for call value and not from both

Reply:
Aside from the generic deposit function, all payable functions in this module take only governance tokens as payment. In any other case, I'd change, but in this specific situation, I believe this is fine.